apps/prairielearn/src/ee/lib/ai-grading.ts

```diff
@@ -344,48 +448,128 @@ export async function aiGrade({
+          messages,
+          model: 'gpt-4o-2024-08-06',
+          user: `course_${course.id}`,
+          response_format: zodResponseFormat(GPTRubricGradeSchema, 'grades'),
+        });
+        try {
+          msg += `Number of tokens used: ${completion.usage ? completion.usage.total_tokens : 0}\n`;
+          const grade_response = completion.choices[0].message;
+          msg += `Raw ChatGPT response:\n${grade_response.content}`;
+          if (grade_response.parsed) {
+            // Only care about the rubric numbers
+            const gpt_rubric_items: number[] = [];
+            for (let i = 0; i < rubric_items.length; i++) {
+              const item = grade_response.parsed.rubric_items[i];
+              if (item.selected) {
+                gpt_rubric_items.push(i);
+              }
+            }
+            const manual_rubric_data = {
+              rubric_id: rubric_items[0].rubric_id,
+              applied_rubric_items: parseRubricItems({
+                rubric_items,
+                gpt_rubric_items,
+              }),
+            };
+            new_rubric_items = await queryRows(
+              sql.select_rubric_for_grading,
+              {
+                assessment_question_id: assessment_question.id,
+              },
+              RubricItemSchema,
+            );
+            // Check if rubric items has been modified
+            validateRubric(rubric_items, new_rubric_items);
+            await manualGrading.updateInstanceQuestionScore(
+              assessment_question.assessment_id,
+              instance_question.id,
+              submission.id,
+              null, // modified_at
+              {
+                feedback: { manual: grade_response.parsed.feedback },
+                manual_rubric_data,
+              },
+              user_id,
+            );
+            msg += `\nAI rubric items: ${gpt_rubric_items.toString()}`;
+          } else if (grade_response.refusal) {
+            job.error(`ERROR AI grading for ${instance_question.id}`);
+            job.error(grade_response.refusal);
+            error_count++;
+          }
+        } catch (err) {
           job.error(`ERROR AI grading for ${instance_question.id}`);
-          job.error(grade_response.refusal);
+          job.error(err);
           error_count++;
         }
-      } catch (err) {
-        job.error(`ERROR AI grading for ${instance_question.id}`);
-        job.error(err);
-        error_count++;
+        job.info(msg);
+        if (warning) {
+          job.warn(warning);
+        }
+        rubric_items = new_rubric_items;
+      } else {
+        const completion = await openai.beta.chat.completions.parse({
+          messages,
+          model: 'gpt-4o-2024-08-06',
+          user: `course_${course.id}`,
+          response_format: zodResponseFormat(GPTGradeSchema, 'grades'),
+        });
+        try {
+          msg += `Number of tokens used: ${completion.usage ? completion.usage.total_tokens : 0}\n`;
```

Comment:
you can just directly call `job.info` each time, no need to accumulate stuff into one big string.

Reply:
I actually think batching is valuable here. I'd highly recommend checking out the relevant part of the server jobs code:

https://github.com/PrairieLearn/PrairieLearn/blob/492975d6878d971ef611396d5b0dc22a69fbbe7a/apps/prairielearn/src/lib/server-jobs.ts#L217-L224

The upshot is that network traffic is more or less $O(n^2)$ with the number of individual written lines, because for reach write, we re-send the _entire_ log up until that point, and there's no internal batching.

If we can update that to only send deltas, or to do internal batching and e.g. update only once a second, we can get the accumulation out of "userspace" here.

Reply:
If we want to batch updates would it be better to actually implement that in the server job code, rather than leaving it to the code that is implementing jobs?

Reply:
Yep! That's what this is about:

> If we can update that _[the core server job code]_ to only send deltas, or to do internal batching and e.g. update only once a second, we can get the accumulation out of "userspace" here.

Reply:
Sounds good. I feel like we could do this pretty simply. Just keep a timestamp called `lastSent` or something, and then when `addToOutput` is called, always add the message to `this.output`, but only make the `socketServer.io` call it if it last happened more than 1 second ago. Then we would also need send all output when the job finished, just be sure we didn't miss the last little bit of it. 

Would that suffice?

Reply:
I've created a PR #10873 to do this.

Reply:
Just made the update